### PR TITLE
Update QdChannel APP after review

### PR DESCRIPTION
**Pull request by mattia-lecci**
Improvements made:
* Added test suite
* Fixed and improved QdChannelModel code
* Added commented-out advanced logs
* Improved example
  * Moved SVD beamforming to a utils script
  * Improved overall quality
  * Reorganized script to highlight the main script
* Added .rst documentation

Questions:
* Should commented-out logs be improved?
* Is the .rst doc ok? How could I visualize it?
* Should we remove the README and point directly to the .rst doc, in order to have a single point of failure?

### CODE REVIEW

**Comment by tommasozugno** [comment]
> Should commented-out logs be improved?

Up to you :)

> Is the .rst doc ok? How could I visualize it?

```
cd ns-3-root/doc/model
make html
```
open index.html on your browser
(make sure to have all the dependencies installed, see the ns-3 installation guide)

> Should we remove the README and point directly to the .rst doc, in order to have a single point of failure?

It would be nice but I don't know if it is possible

File: doc/qd-channel.rst [hunk 1]
```diff
@@ -10,40 +10,52 @@ Example Module Documentation
...
-This is a suggested outline for adding new module documentation to |ns3|.
-See ``src/click/doc/click.rst`` for an example.
-
-The introductory paragraph is for describing what this code is trying to
-model.
-
-For consistency (italicized formatting), please use |ns3| to refer to
-ns-3 in the documentation (and likewise, |ns2| for ns-2).  These macros
-are defined in the file ``replace.txt``.
+Standalone ns-3 application that adds the support for ray-tracing based channel modeling.
```
Comment (mychele) [inline doc/qd-channel.rst]: ns-3 module?

File: doc/qd-channel.rst [hunk 2]
```diff
@@ -10,40 +10,52 @@ Example Module Documentation
...
-Briefly describe the software design of the model and how it fits into 
-the existing ns-3 architecture. 
+The implementation leverages the spectrum implementation of the matrix-based channel model introduced in ns-3.31, described in `this paper <https://arxiv.org/pdf/2002.09341>`_.
 
 Scope and Limitations
 =====================
 
-What can the model do?  What can it not do?  Please use this section to
-describe the scope and limitations of the model.
+The module is able to import channel traces with a given format and create a spectrum channel usable by a generic (or even external) ns-3 module.
```
Comment (mychele) [inline doc/qd-channel.rst]: is able to -> can?

File: doc/qd-channel.rst [hunk 3]
```diff
@@ -10,40 +10,52 @@ Example Module Documentation
...
+The supported format is that of the already-mentioned open-source RT, supporting a number of nodes moving over time, timestep after timestep, with a channel defined between each pair of nodes.
+
+The current approach has the following limitations:
+
+* The folder structure is rigid, an incorrect structure should break the simulation before it even starts, but possible silent fails are not ecluded
```
Comment (mychele) [inline doc/qd-channel.rst]: excluded

File: doc/qd-channel.rst [hunk 4]
```diff
@@ -10,40 +10,52 @@ Example Module Documentation
...
+* M. Lecci, P. Testolina, M. Giordani, M. Polese, T. Ropitault, C. Gentile, N. Varshney, A. Bodi, and M. Zorzi, "Simplified ray tracing for the millimeter wave channel: A performance evaluation", in Proceedings of the Workshop on Information Theory and Applications (ITA), Feb. 2020, San Diego, USA, available here: `arXiv:2002.09179 <https://arxiv.org/abs/2002.09179>`_, DOI: 10.1109/ITA50056.2020.9244950
+* C. Gentile, P. B. Papazian, R. Sun, J. Senic, and J. Wang, “Quasi- Deterministic Channel Model Parameters for a Data Center at 60 GHz,” IEEE Antennas and Wireless Propagation Letters, vol. 17, no. 5, pp. 808–812, May 2018, DOI: 10.1109/LAWP.2018.2817066
+* C. Lai, R. Sun, C. Gentile, P. B. Papazian, J. Wang, and J. Senic, “Methodology for multipath-component tracking in millimeter-wave channel modeling,” IEEE Transactions on Antennas and Propagation, vol. 67, no. 3, pp. 1826–1836, Mar. 2019, DOI: 10.1109/TAP.2018.2888686
+
+
+Please cite our work if you find it interesting!
```
Comment (mychele) [inline doc/qd-channel.rst:61]: All of them? :)

File: doc/qd-channel.rst [hunk 5]
```diff
@@ -55,44 +67,76 @@ into more advanced topics.
...
+.. Helpers
+.. =======
+
+.. What helper API will users typically use?  Describe it here.
 
 Attributes
 ==========
 
-What classes hold attributes, and what are the key ones worth mentioning?
+* Frequency: a read-only attribute needed for compatibility with ``ns3::ThreeGppSpectrumPropagationLossModel``
```
Comment (mychele) [inline doc/qd-channel.rst]: Does a user need to set it?

File: doc/qd-channel.rst [hunk 6]
```diff
@@ -55,44 +67,76 @@ into more advanced topics.
...
 
-Output
-======
+Setting up a scenario
+=====================
+
+This APP is meant to read channel traces generated by the RT mentioned before.
```
Comment (mychele) [inline doc/qd-channel.rst]: module?

File: doc/qd-channel.rst [hunk 7]
```diff
@@ -55,44 +67,76 @@ into more advanced topics.
...
-in additional sections, as needed.
+.. Go into further details (such as using the API outside of the helpers)
+.. in additional sections, as needed.
 
 Examples
 ========
 
-What examples using this new code are available?  Describe them here.
+We currently provide only one example: ``examples/qd-channel-model-example.cc``.
+For more information, please the the source code.
```
Comment (mychele) [inline doc/qd-channel.rst]: see/read

File: model/qd-channel-model.cc [hunk 8]
```diff
@@ -502,7 +512,7 @@ QdChannelModel::GetChannel (Ptr<const MobilityModel> aMob,
   // generate a new channel
   if (notFound || update)
     {
-      NS_LOG_LOGIC ("channelMatrix notFound=" << notFound << " || update=" << update);
+      NS_LOG_DEBUG ("channelMatrix notFound=" << notFound << " || update=" << update);
```
Comment (mychele) [inline model/qd-channel-model.cc]: In general, I would keep logic or info?

**Review by mychele — CHANGES_REQUESTED** [review]

File: doc/qd-channel.rst (hunk 4 above)
Comment (mattia-lecci) [reply doc/qd-channel.rst:61]: what do you suggest? the TWC paper is still under review..

File: doc/qd-channel.rst (hunk 5 above)
Comment (mattia-lecci) [reply doc/qd-channel.rst]: no, it's read-only. should i remove from the doc, then? it's probably unnecessary information..

Comment (mychele) [reply doc/qd-channel.rst]: I would comment better on what you mean with read-only - if you call SetAttribute, what happens?

File: doc/qd-channel.rst (hunk 4 above)
Comment (mychele) [reply doc/qd-channel.rst:61]: I would suggest to suggest the TWC, maybe we can also think about a paper that explicitly describes this implementation.

File: doc/qd-channel.rst (hunk 5 above)
Comment (mattia-lecci) [reply doc/qd-channel.rst]: Done.
FYI: MakeXxxAccessor does not force you to set both a setter and a getter.
You can
1. point directly to a variable
1. point to a getter/setter pair
1. point to a getter only
1. point to a setter only
See `MakeAccessorHelper` in `attribute-accessor-helper.h`

Comment (mychele) [reply doc/qd-channel.rst]: ok - what happens when you call setattribute with an attribute that only has a get?

Comment (mattia-lecci) [reply doc/qd-channel.rst]: btw you get an error

msg="Attribute name=Frequency is not settable for this object: tid=ns3::QdChannelModel", +0.000000000s -1 file=../src/core/model/object-base.cc, line=197
terminate called without an active exception

File: README.md [hunk 9]
```diff
@@ -49,3 +58,14 @@ Configure and build ns-3 from the `ns-3-dev` folder:
 ```
 
 This module does not provide Python bindings at the moment, they will be added in the near future.
+
+### Documentation ###
+
+To compile the documentation, please follow the instructions from the [ns-3 manual](https://www.nsnam.org/docs/manual/html/documentation.html).
+
+1. Istall the dcumentation-specific depdendencies as described in the [ns-3 installation guide](https://www.nsnam.org/wiki/Installation)
```
Comment (mychele) [inline README.md]: Install

Is 1. correct in each bullet?

**Review by mychele — CHANGES_REQUESTED** [review]

File: README.md (hunk 9 above)
Comment (mattia-lecci) [reply README.md]: Aye, it's part of the MD syntax.
Check [here](https://github.com/signetlabdei/qd-channel/blob/updating-app/README.md).

**Review by tommasozugno — APPROVED** [review]

**Review by mychele — APPROVED** [review]

**Pull request merged**